[video_player_avfoundation] iOS platform view support

FirentisTFW opened the pull request:
This PR adds support for platform views on iOS as a way of displaying a video. When creating a video, it's now possible to choose between texture view approach (rendered using `Texture` widget on the Flutter side) and platform view approach (rendered on the native side, using `AVPlayerLayer`).

`FVPVideoPlayer` class now has nothing to do with texture. The texture-related code was moved from it to `FVPTextureBasedVideoPlayer` - a subclass of `FVPVideoPlayer` that adds texture functionality. In the plugin class (`createWithOptions` method) we create either the basic version (for platform view) or the texture subclass (in case of texture approach) based on the parameter passed in from Flutter side.

Platform view is only supported on iOS, no MacOS implementation is added in this PR. The functionality is not yet exposed in the app-facing package (only in example app) - it will be done later, once we add the Android implementation. The PR does not introduce breaking changes, I followed the rule "non-breaking changes, even at the expense of a less-clean API" (`buildViewWithOptions` and `createWithOptions` methods).

Up to this point, the variable naming relied heavily on the texture (we had a lot of `textureId` variables and properties). Since now you can use a platform view instead of a texture view, these variables and parameters were renamed to just `playerId`.

I left some comments in the PR to clarify/discuss some choices.

Resolves [#86613](https://github.com/flutter/flutter/issues/86613) (not fully though, as it is not yet available in the app-facing package).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

FirentisTFW commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:
```diff
@@ -31,6 +39,10 @@ - (FVPDisplayLink *)displayLinkWithRegistrar:(id<FlutterPluginRegistrar>)registr
 
 #pragma mark -
 
+/// The next non-texture player ID, initialized to a high number to avoid collisions with
+/// texture IDs (which are generated separately).
+static int64_t nextNonTexturePlayerId = 1000000;
```
I'm not sure if this is the best approach, but looks like texture ids start from 0 and then get incremented, so this starting point seems to be pretty safe. Alternatively, I assume we could generate random ids for non-texture players

tarrinneal replied:
could use `INFINITY`

FirentisTFW replied:
And you'd suggest to decrement it? Or start with `-INFINITY` and then increment?

tarrinneal replied:
didn't read enough to realize you were incrementing. If that is the case, bigint would make sense for starting. I don't know if we have a standard practice for this though, so maybe @stuartmorgan would have a better solution.

stuartmorgan replied:
Longer term we should fully decouple the ID vended out from texture IDs, and then collisions won't matter; for now to reduce scope, picking something suitably large like `INT_MAX` and decrementing is probably easiest.

FirentisTFW replied:
I used `INT_MAX`.

Should I add a todo for decoupling platform view IDs form texture based IDs? Who I should mention in the todo then? And will we need an issue for it?

FirentisTFW replied:
Since you say it's a relatively minor thing, should it be done in this PR then? How do you see this decoupling?

stuartmorgan replied:
I mean minor in the sense that I don't think it's particularly important, since I think this approach should work fine in practice even though the structure is a bit odd for maintainers to understand.

In code, I would expect that we would make the instance identifiers that are used at the platform interface API layer fully abstract, and track texture IDs separately internally for texture-based instances.

FirentisTFW replied:
Okay, I added a todo for that

FirentisTFW commented:
One PR check is failing, but this is due to changes in platform interface and local dependency overrides. The code can be reviewed already. Once this PR is reviewed, I'll create another one with platform interface changes only (according to `contributing/README.md`).

FirentisTFW commented:
I see two integration tests failing, but I assume this is due to tree-status check failure, since I can see the same tests failing there.

Additionally, I think `testSeekToWhilePausedStartsDisplayLinkTemporarily` (iOS native test) might be flaky. Sometimes it fails (like [here](https://ci.chromium.org/ui/p/flutter/builders/try/Mac_arm64%20macos_platform_tests%20master%20-%20packages/18813/overview)), but when I run the pipeline again (with no additional changes), it passes (or the other way around, first it passes, then it fails). Sometimes it also fails locally for me, but only once every 20-25 test runs (and I could not find a good way to reproduce it).

Anyway, I merged `main` branch, resolved conflicts, and I think the PR can be reviewed. These failing tests hopefully should not be there when the tree is fixed.

tarrinneal commented:
I re-ran the tests for you.

FirentisTFW commented:
Looks like one native test (`testAudioOnlyHLSControls` which was added 3 weeks ago) [was initially failing](https://ci.chromium.org/ui/p/flutter/builders/try/Mac_arm64%20macos_platform_tests%20master%20-%20packages/18926/overview) on MacOS, but it passes now. I could not reproduce the test failure locally, it was always passing, and could not see any issue in the code. So I added some logs to see why it was failing on the CI, but then it passed. I re-ran the tests, and everything passed again. So I removed the logs and force-pushed the original code. And it [passed again](https://ci.chromium.org/ui/p/flutter/builders/try/Mac_arm64%20macos_platform_tests%20master%20-%20packages/19048/overview). Since the issue is not happening anymore, I couldn’t reproduce it, and I already spent almost two days trying to find the cause of it, I’ll publish the PR. Maybe the issue was somehow related to the CI. If it comes back again after I push CR fixes, I’ll have another look at it. Anyway, the PR's ready for review now.

stuartmorgan commented:
That's one of our tests that loads an actual network resource, so it's potentially subject to network-related flake (either at the CI end or the server end).

stuartmorgan commented:
> Additionally, I think `testSeekToWhilePausedStartsDisplayLinkTemporarily` (iOS native test) might be flaky.

I've seen this flake at least once before as well.

stuartmorgan commented on packages/video_player/video_player/example/pubspec.yaml:
```diff
@@ -35,3 +35,15 @@ flutter:
     - assets/bumble_bee_captions.vtt
     - assets/Audio.mp3
+
+# FOR TESTING AND INITIAL REVIEW ONLY. DO NOT MERGE.
```
You can go ahead and revert these tool-generated changes in the packages where nothing else is changing (`video_player`, `video_player_android`, `video_player_web`) now. The tool doesn't know what's going to be changed so it just overrides all references everywhere.

stuartmorgan commented on packages/video_player/video_player_avfoundation/pigeons/messages.dart:
```diff
@@ -17,13 +17,30 @@ import 'package:pigeon/pigeon.dart';
   String? packageName;
   String? formatHint;
   Map<String, String> httpHeaders;
+  PlatformVideoViewType? viewType;
```
In general, something should be nullable only if a clear meaning of `null` can be articulated that is distinct from other meanings. Looking at the native code, it appears that `null` is currently just a second, less intuitive way of writing `PlatformVideoViewType.textureView`, so unless I'm missing something this should be non-nullable at this layer.

FirentisTFW replied:
Yes, it should not be nullable, good catch

stuartmorgan commented on packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:
```diff
@@ -163,6 +173,9 @@ class DataSource {
   final String? package;
+
+  /// The type of view to be used for displaying the video player.
+  final VideoViewType viewType;
```
Why is this part of `DataSource`? How it's displayed isn't related to the video source.

FirentisTFW replied:
The explanation is in the PR description, I did it to avoid breaking changes.
> The PR does not introduce breaking changes, I followed the rule "non-breaking changes, even at the expense of a less-clean API" (buildViewWithOptions method, viewType parameter in DataSource).

Would you suggest a different approach here?

FirentisTFW replied:
Okay, I added `createWithOptions` then and left `DataSource` unchanged

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/include/video_player_avfoundation/FVPVideoPlayerTextureApproach.h:
```diff
@@ -0,0 +1,37 @@
+/// A subclass of FVPVideoPlayer that adds functionality related to texture-based view as a way of
+/// displaying the video in the app. It manages the CALayer associated with the Flutter view,
+/// updates frames, and handles display link callbacks.
+@interface FVPVideoPlayerTextureApproach : FVPVideoPlayer <FlutterTexture>
```
It's not obvious at call sites that this is a player; how about `FVPTextureBasedVideoPlayer`?

FirentisTFW replied:
Good point, renamed to `FVPTextureBasedVideoPlayer`

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/include/video_player_avfoundation/FVPNativeVideoView.h:
```diff
@@ -0,0 +1,32 @@
+#if TARGET_OS_IOS
+@interface FVPNativeVideoView : NSObject <FlutterPlatformView>
+#else
+@interface FVPNativeVideoView : NSWindow
```
Why `NSWindow`? That's a very surprising class for a view to inherit from.

FirentisTFW replied:
Good catch, should be `NSView`

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/include/video_player_avfoundation/FVPNativeVideoView.h:
```diff
@@ -0,0 +1,32 @@
+
+#if TARGET_OS_IOS
+/// Returns the native UIView that displays the video content.
+- (UIView *)view;
```
Methods from implemented protocols are not generally redeclared in Obj-C.

FirentisTFW replied:
Thanks, I took it from [the documentation](https://docs.flutter.dev/platform-integration/ios/platform-views), removed now

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:
```diff
@@ -62,32 +81,44 @@ - (instancetype)initWithAVFactory:(id<FVPAVFactory>)avFactory
+    playerId = [self.registry registerTexture:(FVPVideoPlayerTextureApproach *)player];
+    frameUpdater.textureId = playerId;
+  } else {
+    @synchronized(self) {
```
Why is this `@synchronized`? What other threads are using the ID?

FirentisTFW replied:
Thanks for the explanation, removed `@synchronized` then

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation_ios/FVPNativeVideoView.m:
```diff
@@ -0,0 +1,43 @@
+
+- (void)dealloc {
+  [_view setPlayer:nil];
+  _view = nil;
```
Why is an ivar being nil`d in dealloc?

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation_ios/FVPNativeVideoViewFactory.m:
```diff
@@ -0,0 +1,36 @@
+  NSMutableDictionary<NSNumber *, FVPVideoPlayer *> *_playersById;
+}
+- (instancetype)initWithMessenger:(NSObject<FlutterBinaryMessenger> *)messenger
+                      playersById:(NSMutableDictionary<NSNumber *, FVPVideoPlayer *> *)playersById {
```
It would be cleaner to make this a block that returns a player for an ID, rather than passing in a mutable dictionary and creating the potential for two different classes to be modifying it.

FirentisTFW replied:
Good point, I modified the code like suggested

stuartmorgan commented on packages/video_player/video_player_avfoundation/lib/src/avfoundation_video_player.dart:
```diff
@@ -139,14 +141,45 @@ class AVFoundationVideoPlayer extends VideoPlayerPlatform {
+      VideoViewOptions(
+        playerId: textureId,
+        // Texture view was the only supported view type before
+        // buildViewWithOptions was introduced. We pass it here to maintain
```
I strongly prefer we not put "we" in code comments since it's often ambiguous who "we" is. In this case, I think the second sentence can just be removed entirely.

FirentisTFW commented:
> I've seen this flake at least once before as well.

I replaced a regular assertion with an expectation with a timeout, hope that is okay and fixes the problem.

https://github.com/flutter/packages/pull/8237/commits/9b7f9ad66ea3161970ee6db1c72e5ec4f8c6e70b

ditman commented:
I don't own these packages so removing myself, but this seems like a good entrypoint for an incoming web PR adding a "texture"-like rendering for the video player there. [PR](https://github.com/flutter/packages/pull/6286) (it got closed though, so I'm not sure if it'll ever get updated to leverage this).

FirentisTFW commented on packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:
```diff
@@ -476,3 +501,35 @@ class VideoPlayerWebOptionsControls {
+  /// Constructs an instance of [VideoCreationOptions].
+  const VideoCreationOptions({
+    required this.dataSource,
+    required this.viewType,
```
@stuartmorgan Do you think we should put `VideoViewType.textureView` as a default here?

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:
```diff
@@ -31,6 +39,11 @@ - (FVPDisplayLink *)displayLinkWithRegistrar:(id<FlutterPluginRegistrar>)registr
+// TODO(stuartmorgan): Decouple IDs for platform views and texture views.
+/// The next non-texture player ID, initialized to a high number to avoid collisions with
+/// texture IDs (which are generated separately).
+static int64_t nextNonTexturePlayerId = INT_MAX;
```
This can just be an ivar; we don't need IDs to be unique across engine instances.

FirentisTFW replied:
Right, changed to ivar

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:
```diff
@@ -62,32 +84,42 @@ - (instancetype)initWithAVFactory:(id<FVPAVFactory>)avFactory
-  int64_t textureId = [self.registry registerTexture:player];
-  frameUpdater.textureId = textureId;
+  BOOL textureBased =
+      frameUpdater != nil && [player isKindOfClass:[FVPTextureBasedVideoPlayer class]];
```
Why is `frameUpdater != nil` part of this check? If someone passed in a texture-based player without a frame updater (which would be a bug), it would still be texture-based.

FirentisTFW replied:
You're right, removed this part

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation_ios/FVPNativeVideoViewFactory.m:
```diff
@@ -0,0 +1,38 @@
+  return FVPGetMessagesCodec();
+}
+
+@end
```
Nit: please add the missing newline.

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation_ios/FVPNativeVideoViewFactory.m:
```diff
@@ -0,0 +1,38 @@
+
+@implementation FVPNativeVideoViewFactory {
+  NSObject<FlutterBinaryMessenger> *_messenger;
+  FVPVideoPlayer * (^_playerByIdProvider)(NSNumber *);
```
The preferred repository style for new code is to use private properties and use `self.` access outside of init/dealloc.

FirentisTFW commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPTextureBasedVideoPlayer.m:
```diff
@@ -0,0 +1,180 @@
+  AVPlayerItem *item = [AVPlayerItem playerItemWithAsset:urlAsset];
+  return [self initWithPlayerItem:item
+                     frameUpdater:(FVPFrameUpdater *)frameUpdater
+                      displayLink:(FVPDisplayLink *)displayLink
```
Looks like displayLink cast was actually there before. I removed them, thanks

FirentisTFW commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPTextureBasedVideoPlayer.m:
```diff
@@ -0,0 +1,180 @@
+  _displayLink.running = self.isPlaying || self.waitingForFrame;
+}
+
+- (void)seekTo:(int64_t)location completionHandler:(void (^)(BOOL))completionHandler {
```
Sure, I had a fixme comment there to clean it up, but then looks like I just removed it

FirentisTFW commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPTextureBasedVideoPlayer.m:
```diff
@@ -0,0 +1,180 @@
+  _displayLink.running = YES;
+}
+
+- (CVPixelBufferRef)copyPixelBuffer {
```
Added pragma sections for `FlutterTexture`, `Overrides` and `Private methods`

FirentisTFW commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayer.m:
```diff
@@ -128,22 +87,12 @@ - (instancetype)initWithPlayerItem:(AVPlayerItem *)item
-  // invisible AVPlayerLayer is used to overwrite the protection of pixel buffers in those streams
-  // for issue #1, and restore the correct width and height for issue #2.
-  _playerLayer = [AVPlayerLayer playerLayerWithPlayer:_player];
-  [self.flutterViewLayer addSublayer:_playerLayer];
```
Yes, I checked that on iOS 16 and it looks good. When I comment these two lines out, the aspect ration is broken again, so I assume the fix is still working, at least for the second issue. The first one (blank screen) I couldn't reproduce anymore (on iOS 16), I guess it's working fine, we also have a unit test for it.

stuartmorgan commented on packages/video_player/video_player_platform_interface/lib/video_player_platform_interface.dart:
```diff
@@ -48,11 +48,19 @@ abstract class VideoPlayerPlatform extends PlatformInterface {
 
+  /// Creates an instance of a video player based on creation options
+  /// and returns its playerId. If view type is [VideoViewType.textureView],
+  /// the playerId is also the texture id.
```
Looking at just the platform interface part again, I don't think we should encode this into the interface, as it will make it harder for the higher-level code to untangle the now-ideally-abstract player ID from the texture ID, which has a very specific meaning.

There's also a design issue here where both `createWithOptions` and `buildViewWithOptions` are being passed a view type, but once `createWithOptions` has been called, there is only one valid type that can be passed to `buildViewWithOptions` for that player ID, so the calling code is being forced to maintain view type state.

Couldn't `VideoViewOptions` only contain the player ID for now, and the state of how to map player ID to implementation details like the texture ID (if any) be maintained in the platform implementation, which is already stateful?

FirentisTFW replied:
I did as suggested. Now the responsibility of maintaining view type state for a given player relies on the platform implementation.

The player ID as viewed by the app facing package should now more abstract, but currently we ended up in a very ugly state where most of the methods still use `textureId` for parameter names (which now should be renamed to `playerId`). I suppose that we could fix it all up when creating platform interface PR?

In case of texture view, the player ID will still be the same as texture ID, we would just not explicitly reveal that. If we wanted to change this and make the player ID fully decoupled from the texture ID, that would probably require some significant changes - either moving the player ID generation process to the Dart side or changing the pigeon interface so that it returns player ID as well as texture ID (which can be null).

stuartmorgan replied:
> In case of texture view, the player ID will still be the same as texture ID, we would just not explicitly reveal that.

That's fine for now as an implementation detail, I just don't want that to become something we are stuck with indefinitely because it's part of the API contract of the new methods.

FirentisTFW commented on packages/video_player/video_player_avfoundation/test/avfoundation_video_player_test.dart:
```diff
@@ -188,6 +188,126 @@ void main() {
+      expect(textureId, 3);
+    });
+
+    test('createWithOptions with incorrect asset throws exception', () async {
+      try {
+        await player.createWithOptions(
+          VideoCreationOptions(
+            dataSource: DataSource(
+              sourceType: DataSourceType.asset,
+              asset: '/path/to/incorrect_asset',
+            ),
+            viewType: VideoViewType.textureView,
+          ),
+        );
+        fail('should throw PlatformException');
+      } catch (e) {
+        expect(e, isException);
+      }
+    });
```
I copied this test from the one above (for `create` method) which was added in [this PR](https://github.com/flutter/packages/pull/4318), but I don't think it checks anything.

I assume the call to `createWithOptions` (or `create` in the original test) should fail, then we should catch it, and verify the exception. But looks like what actually happens is that the `fail` call throws, we just catch it and verify that it is exception (which it is because it's a `TestFailure`).

Do we need this test? What value does it bring?

stuartmorgan replied:
> But looks like what actually happens is that the `fail` call throws, we just catch it and verify that it is exception (which it is because it's a `TestFailure`).

Sounds like the test is broken; the preferred way to write it would be with a `throwsA` matcher on the call, instead of a manual try/catch.

FirentisTFW replied:
Does this test make sense here in the first place? [The exception](https://github.com/flutter/packages/pull/4318/files#diff-85f606af958fc0025baafe608f631ef8622a8f35e583c1e4f3b5f99cfdc659de) it is supposed to check comes from the native side.

stuartmorgan replied:
Ugh, I see; sorry, I hadn't looked at the initial PR. You are correct, this test makes no sense here. It should have been either a native unit test, or a Dart integration test, since a Dart unit test can't possibly exercise that code. Unfortunately that was missed in review.

FirentisTFW replied:
I replaced it with a native unit test.

stuartmorgan reviewed (commented):
I didn't fully re-review, but I looked at the platform interface part and skimmed how that affected the implementation. It looks like this approach does what we need, and will be clearly applicable to other platforms (and if ever necessary, added view types) so I think we're good to split that piece out to its own PR now, so that it can be landed and unblock Android work, while the iOS review proceeds.

FirentisTFW commented on packages/video_player/video_player_avfoundation/lib/src/avfoundation_video_player.dart:
```diff
@@ -15,6 +15,11 @@ import 'messages.g.dart';
+  /// A map that associates player ID with a view type.
+  /// This is used to determine which view type to use when building a view.
+  @visibleForTesting
+  final Map<int, VideoViewType> playerViewTypes = <int, VideoViewType>{};
```
Good point, I did as suggested.

hellohuanlin commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPTextureBasedVideoPlayer.m:
```diff
@@ -0,0 +1,178 @@
+
+#import "FVPTextureBasedVideoPlayer_Test.h"
+
+@interface FVPTextureBasedVideoPlayer ()
```
What's the corresponding type for platform view based implementation? is there a `FVPPlatformViewBasedVideoPlayer`?

FirentisTFW replied:
Actually, there's no additional type for platform view. It uses regular `FVPVideoPlayer` class as it has all functionalities needed for a platform view to work. Would you see that differently?

hellohuanlin replied:
Are there any functionalities (properties or methods) unique to the platform view based implementation, that is not applicable to texture based implementation?

FirentisTFW replied:
Nope, platform view player uses a subset of texture player features. That's why I did it this way - `FVPVideoPlayer` as a base class that could be also used directly in case of platform views

hellohuanlin replied:
Gotcha. Can you add some comments in `FVPVideoPlayer` explaining this?

FirentisTFW replied:
Yes, updated the class doc

hellohuanlin commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:
```diff
@@ -158,69 +193,84 @@ - (nullable NSNumber *)createWithOptions:(nonnull FVPCreationOptions *)options
+  if ([player isKindOfClass:[FVPTextureBasedVideoPlayer class]]) {
+    [self.registry unregisterTexture:playerId];
+  }
+  [self.playersById removeObjectForKey:playerKey];
```
Can we have 2 types of video players both implementing the same API and rely on polymorphism?

FirentisTFW replied:
Hmm, isn't that what we currently have? Right now the structure is as follows:
- `FVPVideoPlayer` (contains the basic functionality and can be used directly for platform views),
- `FVPTextureBasedVideoPlayer` (extends `FVPVideoPlayer` adding texture-specific functionalities).
Would you change something here?

Or, if you're talking about this piece of code specifically, we could also call `unregisterTexture` in `[player dipose]`, then we'd need to pass `FlutterTextureRegistry` there (or an additional callback, like `onDisposed`). Shall I do that?

hellohuanlin replied:
I was thinking about ways to remove this kind of logic which happens a lot in this file: 

```
func doSomething() {
  if texture based {
    do A
  } else {
    do B
  }
}
```

I was suggesting having 2 players, each encapsulating its own logic, so that we don't mix the 2 logic in the same file: 
```
class TextureBasedPlayer {
  func doSomething() {
    do A
  }
}
class PlatformBasedPlayer {
  func doSomething() {
    do B
  }
}
```

then in this function we can just do: 
```
func doSomething() {
  player.doSomething()
}
```

FirentisTFW replied:
I understand, I think this is what we're currently doing though.

I general, that's right, we do have some checks in this file. However, most of them are in `createWithOptions` method, and these checks are just to properly initialize the player, I'm not sure if we can move them out of there (though probably the method itself can be refactored and/or split up, I'll have another look into it tomorrow). If we cross out those, we're left with two checks:
- Here in `disposePlayer`, which I think we can actually get rid of if we pass `onDisposed` callback when initializing the player.
- In `onPlayerSetup` where we obtain a player ID for the player that was just created

FirentisTFW replied:
I cleaned up a little bit the code related to player initialization (`createWithOptions` method) and got rid of the type check in `dispose`

hellohuanlin commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:
```diff
@@ -132,95 +163,130 @@ - (void)initialize:(FlutterError *__autoreleasing *)error {
+  }
+}
+
+- (nullable FVPTextureBasedVideoPlayer *)createTextureBasedPlayerFromOptions:
```
nit: *With*Options

hellohuanlin commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:
```diff
@@ -132,95 +163,130 @@ - (void)initialize:(FlutterError *__autoreleasing *)error {
 
-  FVPVideoPlayer *player;
+  void (^onDisposed)(int64_t) = ^(int64_t textureId) {
+    [self.registry unregisterTexture:textureId];
```
weak self

FirentisTFW replied:
Right, fixed

hellohuanlin commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:
```diff
@@ -132,95 +163,130 @@ - (void)initialize:(FlutterError *__autoreleasing *)error {
+  return nil;
+}
+
+- (nullable FVPVideoPlayer *)createRegularPlayerFromOptions:(nonnull FVPCreationOptions *)options {
```
what's a "regular" player

FirentisTFW replied:
I updated the method name to `createPlatformViewPlayerWithOptions` and left a comment that using `FVPVideoPlayer` is sufficient there

FirentisTFW commented:
@stuartmorgan I [added a fallback](https://github.com/flutter/packages/pull/8237/commits/417e7716df27150aa5c3cc17af61215a341517ff) to always use texture view on MacOS, even if platform view is requested. I'm not sure how to best document it. For sure I'll put the information about it in app-facing package (when we pass the view type to controller), but I'm thinking if we should also describe it better here, in the platform implementation. What are your thoughts on that?

stuartmorgan commented:
> I'm thinking if we should also describe it better here, in the platform implementation. What are your thoughts on that?

We mostly don't expect clients to interact with the video player implementation packages directly, but it wouldn't hurt to add a `## Platform limitations` section to the README for this package that mentions the lack of support.

hellohuanlin commented on packages/video_player/video_player_avfoundation/README.md:
```diff
@@ -13,3 +13,7 @@ should add it to your `pubspec.yaml` as usual.
+
+## Platform limitations
+
+On macOS, the plugin does not support platform views. Instead, a texture view is always used to display the video player, even if `VideoViewType.platformView` is specified as a parameter.
```
This should be documented in API as well (if it's not already so)

FirentisTFW replied:
I added a comment in the implementation [here](https://github.com/flutter/packages/pull/8237/commits/417e7716df27150aa5c3cc17af61215a341517ff#diff-1a110fd39c95303a6bd97033203f739866aab30518994ac55bf134e7d9f60e03R56-R59). Then based on the discussion with Stuart ([question](https://github.com/flutter/packages/pull/8237#issuecomment-2622020563) and [reply](https://github.com/flutter/packages/pull/8237#issuecomment-2622706965)) I mentioned it in the README. Would you suggest to add a dartdoc on `createWithOptions` method where this fallback is implemented?

stuartmorgan replied:
> since that's where developers use the api

~0 developers outside of the Flutter team will use `createWithOptions`. It's a platform interface API used by the app-facing package.

FirentisTFW replied:
I'm leaving it as is then

hellohuanlin commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/include/video_player_avfoundation/FVPVideoPlayer.h:
```diff
@@ -77,10 +65,34 @@ NS_ASSUME_NONNULL_BEGIN
+/// The plugin registrar, to obtain view information from.
+@property(nonatomic, readonly) NSObject<FlutterPluginRegistrar> *registrar;
+/// The CALayer associated with the Flutter view this plugin is associated with, if any.
+@property(nonatomic, readonly, nullable) CALayer *flutterViewLayer;
```
is this CALayer only for platform view based implementation, but not applicable to texture based implementation?

FirentisTFW replied:
It's actually only needed in texture based implementation ([this comment](https://github.com/flutter/packages/pull/8237/files#diff-79247173115dd0e1159d9abc3dfbbfb6666258e7c3d7a0814b40fe5d751614b5R80-R86) says it fixes two bugs there). It should not be in `FVPVideoPlayer` anymore, looks like it got there when I was resolving conflicts. Thanks, removed it now

hellohuanlin commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/include/video_player_avfoundation/FVPNativeVideoView.h:
```diff
@@ -0,0 +1,27 @@
+/// A class used to create a native video view that can be embedded in a Flutter app.
+/// This class wraps an AVPlayer instance and displays its video content.
+#if TARGET_OS_IOS
+@interface FVPNativeVideoView : NSObject <FlutterPlatformView>
```
Could you point me to where we add the view to the view hierarchy?

FirentisTFW replied:
[Here](https://github.com/flutter/packages/pull/8237/files#diff-0da911603637e04d371c207214dc218fb75a1a2d7520f2120ba5c1b3348b25dd) is the native platform view factory implementation where we create instances of `FVPNativeVideoView`, and [here](https://github.com/flutter/packages/pull/8237/files#diff-1a110fd39c95303a6bd97033203f739866aab30518994ac55bf134e7d9f60e03R189-R216) is where we create the view on the Flutter side.

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/include/video_player_avfoundation/FVPVideoPlayer.h:
```diff
@@ -77,10 +68,32 @@ NS_ASSUME_NONNULL_BEGIN
 - (void)seekTo:(int64_t)location completionHandler:(void (^_Nullable)(BOOL))completionHandler;
+@end
+
+@interface FVPVideoPlayer ()
```
Why is there a private category in the public header? If these should be public, they can just be part of the main interface.

stuartmorgan replied:
Although it's not clear to me why most of this would need to be public. Is this for subclassing? If so, that can be in an `..._Internal.h` file.

FirentisTFW replied:
Yes, it's for subclassing. I wasn't aware of `..._Internal.h` pattern, I thought that having a class extension here is the common practice. I moved it to `FVPVideoPlayer_Internal.h` as suggested

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/include/video_player_avfoundation/FVPVideoPlayer.h:
```diff
@@ -77,10 +68,32 @@ NS_ASSUME_NONNULL_BEGIN
-/// play/pause state.
-- (void)expectFrame;
+/// Updates the playing state of the video player.
+- (void)updatePlayingState;
```
Why does this need to be part of the public interface? It should be an implementation detail.

FirentisTFW replied:
It was for subclassing, answered [here](https://github.com/flutter/packages/pull/8237#discussion_r1943269864).

hellohuanlin reviewed (approved):
thanks for making the changes

stuartmorgan commented on packages/video_player/video_player_avfoundation/pubspec.yaml:
```diff
@@ -24,7 +24,7 @@ flutter:
   flutter:
     sdk: flutter
-  video_player_platform_interface: ">=6.1.0 <7.0.0"
+  video_player_platform_interface: ">=6.3.0 <7.0.0"
```
This can just be `^6.3.0`.

stuartmorgan commented on packages/video_player/video_player_avfoundation/example/pubspec.yaml:
```diff
@@ -16,7 +16,7 @@ dependencies:
     # the parent directory to use the current plugin's version.
     path: ../
-  video_player_platform_interface: ">=6.1.0 <7.0.0"
+  video_player_platform_interface: ">=6.3.0 <7.0.0"
```
`^6.3.0`

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPTextureBasedVideoPlayer.m:
```diff
@@ -0,0 +1,205 @@
+  if (!path) {
+    path = [NSURL URLWithString:asset relativeToURL:NSBundle.mainBundle.bundleURL].path;
+  }
+#endif
```
Asset key -> path logic should be extracted to a shared helper function, rather than duplicated.

FirentisTFW replied:
Good catch, missed that. Extracted to a helper method

tarrinneal commented:
I'll get to this review asap. Thanks for your patience.

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/include/video_player_avfoundation/FVPNativeVideoViewFactory.h:
```diff
@@ -0,0 +1,24 @@
+/// Initializes a new instance of FVPNativeVideoViewFactory with the given messenger and
+/// a block that provides video players associated with their IDs.
+- (instancetype)initWithMessenger:(NSObject<FlutterBinaryMessenger> *)messenger
+               playerByIdProvider:(FVPVideoPlayer * (^)(NSNumber *))playerByIdProvider;
```
Nit: Id -> Identifier, per standard Obj-C naming.

stuartmorgan replied:
Ideally we would do it everywhere, but that's complicated by the fact that we don't have an annotation in Pigeon for renaming class members, params, etc., and `playerId` *is* idiomatic for Dart, so some amount of `playerId` will filter into the Obj-C code. I didn't audit for how much inconsistency we would get if we renamed all the Obj-C members and left just the Pigeon-generated code as using the non-idiomatic naming.

If the inconsistency is fairly limited that changing it everywhere would be preferred, as `Id` is very much not standard Obj-C (`Identifier` is dominant in Apple SDKs, occasionally they use `ID`). But if changing variables would add a lot of inconsistency I'd settle for just making method signatures use standard naming (thus flagging just this one).

> I had a quick look into other packages and looks like we use both `Id` and `Identifier` there.

The nature of this repo is unfortunately that because there's so much context switching between languages—even in a single PR—things that break language-specific naming idioms are frequently missed in review. I still try though 🙂

FirentisTFW replied:
Okay, thanks for the detailed explanation. I'll see what I can do with renaming the params then. I'll let you know when I'm finished with this

FirentisTFW commented on packages/video_player/video_player_avfoundation/README.md:
```diff
@@ -13,3 +13,7 @@ should add it to your `pubspec.yaml` as usual.
+
+## Platform limitations
+
+On macOS, the plugin does not support platform views. Instead, a texture view is always used to display the video player, even if `VideoViewType.platformView` is specified as a parameter.
```
Right, good point

FirentisTFW commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/include/video_player_avfoundation/FVPNativeVideoView.h:
```diff
@@ -0,0 +1,27 @@
+#import <Flutter/Flutter.h>
+#endif
+
+#import <AVFoundation/AVFoundation.h>
```
Thanks, moved conditional imports to the end

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPTextureBasedVideoPlayer.m:
```diff
@@ -0,0 +1,197 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "./include/video_player_avfoundation/FVPTextureBasedVideoPlayer_Test.h"
```
This should be:

```objc
#import "./include/video_player_avfoundation/FVPTextureBasedVideoPlayer.h"
#import "./include/video_player_avfoundation/FVPTextureBasedVideoPlayer_Test.h"
```

FirentisTFW replied:
Thanks, I was wondering about that part

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:
```diff
@@ -62,32 +81,44 @@ - (instancetype)initWithAVFactory:(id<FVPAVFactory>)avFactory
+- (int64_t)onPlayerSetup:(FVPVideoPlayer *)player {
+  BOOL textureBased = [player isKindOfClass:[FVPTextureBasedVideoPlayer class]];
+  int64_t playerId;
+  if (textureBased) {
```
Optional: an alternate way to structure this would be:

```objc
FVPTextureBasedVideoPlayer* textureBasedPlayer = [player isKindOfClass:[FVPTextureBasedVideoPlayer class]] ?
    (FVPTextureBasedVideoPlayer *)player : nil;

if (textureBasedPlayer) {
  playerId = [self.registry registerTexture:textureBasedPlayer];
  ...
}
...
[textureBasedPlayer expectFrame];
```

Then the only casting is happening in the same statement with the check that makes the cast safe, so there's less room for error in future refactoring.

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:
```diff
@@ -132,95 +163,133 @@ - (void)initialize:(FlutterError *__autoreleasing *)error {
+      player = [self createTexturePlayerWithOptions:options];
+    } else {
+      player = [self createPlatformViewPlayerWithOptions:options];
+    }
```
Optional:

```objc
FVPVideoPlayer *player = textureBased
    ? [self createTexturePlayerWithOptions:options]
    : [self createPlatformViewPlayerWithOptions:options];
```

(This keeps the declaration and initialization in the same statement, making it clearer that it's always initialized.)

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:
```diff
@@ -132,95 +163,133 @@ - (void)initialize:(FlutterError *__autoreleasing *)error {
+  }
+}
+
+- (nullable FVPTextureBasedVideoPlayer *)createTexturePlayerWithOptions:
```
Nit: the naming of these two methods would be more idiomatic if you remove the `create`. Convenience constructors in Obj-C generally just start with the name of the thing they create.

FirentisTFW replied:
Thanks, wasn't aware of that. Fixed now!

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:
```diff
@@ -132,95 +163,133 @@ - (void)initialize:(FlutterError *__autoreleasing *)error {
-  [self.playersByTextureId removeObjectForKey:playerKey];
-  if (!player.disposed) {
-    [player dispose];
+- (NSString *)prepareAssetPathFromCreationOptions:(nonnull FVPCreationOptions *)options {
```
Nit: `assetPathFromCreationOptions:`

(Similar reason here; since the method is intended to create and return something, the name should generally start by naming that thing, without a verb.)

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/include/video_player_avfoundation/FVPNativeVideoViewFactory.h:
```diff
@@ -0,0 +1,24 @@
+
+// Platform views are only supported on iOS as of now. Ifdef is used to avoid compilation errors.
+
+#import <TargetConditionals.h>
```
This file should import `Foundation` given the types it's using, at which point you won't need this.

stuartmorgan commented on packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/include/video_player_avfoundation/FVPVideoPlayer_Internal.h:
```diff
@@ -0,0 +1,47 @@
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FVPVideoPlayer ()
```
```objc
/// Interface intended for use by subclasses, but not other callers.
```

stuartmorgan reviewed (approved):
LGTM with minor comments (none that need another round of review, and feel free to ignore the optional ones if you disagree with them). For the `Id`/`Identifier` thing, you can use your judgement about where to stop renaming.

If there's anything you'd specifically like me to re-review let me know and I'll take a look, but otherwise feel free to land once the comments are addressed!

FirentisTFW commented:
@stuartmorgan @hellohuanlin Thanks for the review, I addressed all of your comments.

For `Id`/`Identifier` - I renamed the `Id` to `Identifier` where possible. It worked great, now `Id` is mostly left in pigeon-generated files.
I also applied optional suggestions from the last review.

Thanks for the iteration here, I'm merging it now.
